ikev2: detect and resolve simultaneous IKE_AUTH exchanges

The-Mule opened the pull request:
When processing an IKE_AUTH request as responder, detect if there's a simultaneous IKE SA on the same permanent connection where we are the initiator with an outstanding IKE_AUTH request.

If this is detected, decide which IKE SA will be rejected and which one will continue based on this logic:

 - If the simultaneous (initiator) IKE SA already completed IKE_AUTH within the last second, keep it and reject the current (responder) IKE_AUTH request.

 - Otherwise, compare the initiator nonces from both exchanges, keep IKE SA with the higher nonce value and reject the other.

If rejecting responder's IKE SA we send AUTHENTICATION_FAILED. Otherwise, we terminate the initiator's IKE SA locally.

Note:

 - AUTHENTICATION_FAILED is sent (instead of TEMPORARY_FAILURE) because it is fatal and deletes entire IKE SA. This ensures clean resolution: one IKE SA deleted, one continues and revival establishes a fresh connection if needed.

 - This detection is not enabled by default but it can be enabled by setting connection option reject-simultaneous-ike-auth=yes.

Resolves: #2184 #2445

The-Mule commented:
@cagney @ueno, can you take a peek if this is a way to go?

cagney commented:
Rejecting the IKE SA for the current message seems right.

Sending back TEMPORARY_FAILURE before the IKE SA is established is off-script.  Since it isn't fatal, the initiator is presumably expected to (after a pause) initiate a new exchange (here IKE_AUTH) with a new Message ID.

We need consider conn-switching.  That's when the IKE_SA_INIT responder picks the wrong connection.  It's only once the IKE SA is authenticated that we can trust the connection is correct.  Same thing goes for the Child SA, only once the Traffic Selectors are processed can the Child SA's connection be trusted.

What we could try is this code under a very narrow set of circumstances.  Namely when the connection can't switch.

BTW, is the intent for the check to happen at the start of the second packet, where the connection is secured but not authenticated,  or IKE_AUTH (which might be after several IKE_INTERMEDIATE exchanges)?

cagney commented:
> > For instance, an established wedged competing_ike would stop this exchange until the other IKE SA times out.
> 
> Can you elaborate more on this? If I get it right then this happens if and only if `competing_ike` has IKE_AUTH request outstanding but it's a sort of lost and hence we can't really expect IKE_AUTH response. But I am wondering if there is anything we can do in this situation...?

When things die, there's a good chance that an established competing_ike has an outstanding liveness probe.
Limiting things to larval (!established) IKE SAs would avoid that.

mamtagambhir commented:
> Pushed v2 patch. It seems to be passing my testing but definitely more testing is needed.

I can help test the fix if its ready, Can I patch it on top of released 5.3?

The-Mule commented:
The last version covers also connection switch during TS negotiation. I tested behavior without connection switching so far and it looks like it is working unless I missed something (also rebased on top of current main branch).

The-Mule commented on programs/pluto/ikev2_ts.c:
```diff
@@ -1655,6 +1656,12 @@ bool process_v2TS_request_payloads(struct ike_sa *ike,
+
+			if (has_outstanding_ike_auth_request(best.connection, ike, md)) {
+				llog_ts(child, &tsps, "best connection matching TS has IKE_AUTH request outstanding");
+				return false;
```
@cagney Hm, this actually leads to childless IKE SA but is there a way to drop IKE SA at this point?

The-Mule committed:
ikev2: detect and resolve simultaneous IKE_AUTH exchanges

When processing an IKE_AUTH request as responder, detect if there's
a simultaneous IKE SA on the same permanent connection where we are
the initiator with an outstanding IKE_AUTH request.

If this is detected, keep only one IKE SA based on this logic:

  - If the simultaneous (initiator) IKE SA already completed IKE_AUTH
    within the last second, keep it and reject the current (responder)
    IKE_AUTH request.

  - Otherwise, compare the initiator nonces from both exchanges and
    keep the IKE SA with the higher nonce value.

If rejecting responder's IKE SA we send AUTHENTICATION_FAILED. Otherwise,
we terminate the initiator's IKE SA locally. This is enabled by default
and can be disabled by connection option reject-simultaneous-ike-auth=no.

Signed-off-by: Ondrej Moris <omoris@redhat.com>
Signed-off-by: Daiki Ueno <dueno@redhat.com>

The-Mule commented:
@cagney It turns out that in theory this approach might lead to repeated AUTHENTICATION_FAILED responses - when IKE AUTH requests cross and both sides get AUTHENTICATION_FAILED then it is possible that they start their new IKE AUTH exchange again at the same time and again and again. It is a fairly low chance since network jitter will most likely prevent this. Hence you can see it rather on a test machine with many network namespaces that in real world. 

But I there might actually be a better way. Perhaps it would be better not to drop both IKE SA but keep the one with higher SPI continue and drop the one with lower SPI. This would mean that if the other side IKE AUTH request has IKE SA with lower SPI, then responder would reply with AUTHENTICTION_FAILED. Otherwise resnpoder would reply with IKE AUTH response. The same responder will eventually get AUTHENTICATION_FAILED from the intiator and drop its own IKE SA in favour of initiator's one. This approach would also prevent aforemention cycle.

I did not think through all the details but the changes should be rather minimal.

`has_outstanding_ike_auth_request()` will change to `get_sa_with_outstanding_ike_auth_request()` so that it actually returns simultaneous IKE SA and then instead of unconditionally replying AUTHENTICATION_FAILED we'll do:

```
struct ike_sa *simultaneous_ike = get_sa_with_outstanding_ike_auth_request(c, ike, md);
if (simultaneous != NULL) {
  // Compare our initiated SPI vs their initiated SPI (from message)
  int cmp = ...
  
if (cmp > 0) {
  // Our IKE SA with oustanding IKE AUTH request has SPI higher, delete this IKE SA, keep the simultaneous_ike
  record_v2N_response(ike->sa.logger, ike, md, v2N_AUTHENTICATION_FAILED, empty_shunk, ENCRYPTED_PAYLOAD);
  return STF_FATAL;
} else {
  // IKE SA for the current IKE AUTH request has SPI higher, continue with IKE AUTH reply and drop the other one
  delete_ike_sa(&simultaneous_ike);
  }
}
```

This approach is also a bit less strict than the previous one (where we dropped ball on both sides). If the other side gets AUTHENTICATION_FAILED it is supposed to retry since we are dealing with permanent connections but that is not a problem since at that point there will be Child SA already created.

FYI, already merged part of the PR is still valid. 

(cc @ueno)

ueno commented:
@The-Mule that sounds a bit tricky but I guest it would work; I've implemented it as https://github.com/ueno/libreswan/commit/38ecdd4be799c753e3d28e10b8041a1e2e65f806.

On the other hand, if we manually delete the simultaneous_ike (initiated by the other), I wonder if it would be possible to just accept the peer's request and withdraw ours, without sending AUTHENTICATION_FAILURE?

The-Mule committed:
impair: add impair option for setting nonces

New impair options 'impair_initiator_nonce' and 'impair_responder_nonce'
allow setting all 32 bytes of nonce to specified value, e.g. to set high
nonce value one can use 'impair ike_initiator_nonce:0xff'. This might be
useful for testing.

Signed-off-by: Ondrej Moris <omoris@redhat.com>

cagney commented:
@The-Mule it's meant to be, adding tag

cagney commented:
First run came back with:
```
linux-audit-01-ok-ikev1
interop-ikev2-strongswan-35-rekey-pfs
crossing-streams-11-ikev2-permanent-ike-auth-east
crossing-streams-24-ikev2-delete-connswitch-github-2101
crossing-streams-11-ikev2-permanent-ike-auth-west
crossing-streams-20-peer-restarts-ondemand-ike-auth
```
but the recheck narrowed it down to just crossing streams:

The first group are just mainline changes:
```
--- MAIN/testing/pluto/crossing-streams-11-ikev2-permanent-ike-auth-east/west.console.txt
+++ OUTPUT/testing/pluto/crossing-streams-11-ikev2-permanent-ike-auth-east/west.console.txt
@@ -45,7 +45,7 @@
 "east-west" #2: sent IKE_SA_INIT request to 192.1.2.23:UDP/500
 west #
  ../../guestbin/wait-for-inbound.sh 3
-IMPAIR: packet fr/om 192.1.2.23:500: blocking inbound message 3
+IMPAIR: packet from 192.1.2.23:500: blocking inbound message 3
 west #
  # Process IKE_SA_INIT response from EAST
 west #
@@ -58,7 +58,7 @@
  # Wait for WEST sent IKE_AUTH request
 west #
  ../../guestbin/wait-for-pluto.sh '#2: sent IKE_AUTH request'
-"east-west" #2: sent IKE_AUTH request to 192.1.2.23:UDP/500 with shared-key-mac and FQDN '@west'; Child SA #3 {ESP <0xESPESP}
+"east-west" #2: sent IKE_AUTH request to 192.1.2.23:UDP/500 with shared-key-mac and FQDN '@west'; Child SA #3 {ESP <0xESPESP} [192.1.2.45/32===192.1.2.23/32]
 west #
  ../../guestbin/wait-for-inbound.sh 4
 IMPAIR: packet from 192.1.2.23:500: blocking inbound message 4
--- MAIN/testing/pluto/crossing-streams-11-ikev2-permanent-ike-auth-west/west.console.txt
+++ OUTPUT/testing/pluto/crossing-streams-11-ikev2-permanent-ike-auth-west/west.console.txt
@@ -58,7 +58,7 @@
  # Wait for WEST sent IKE_AUTH request
 west #
  ../../guestbin/wait-for-pluto.sh '#2: sent IKE_AUTH request'
-"east-west" #2: sent IKE_AUTH request to 192.1.2.23:UDP/500 with shared-key-mac and FQDN '@west'; Child SA #3 {ESP <0xESPESP}
+"east-west" #2: sent IKE_AUTH request to 192.1.2.23:UDP/500 with shared-key-mac and FQDN '@west'; Child SA #3 {ESP <0xESPESP} [192.1.2.45/32===192.1.2.23/32]
 west #
  ../../guestbin/wait-for-inbound.sh 4
 IMPAIR: packet from 192.1.2.23:500: blocking inbound message 4
```
we can see the TS as they go out - hoot! Although `fr/om` does have me puzzled.

The next might be a sign of the new code, but I don't see why both didn't establish:
```
--- MAIN/testing/pluto/crossing-streams-24-ikev2-delete-connswitch-github-2101/east.console.txt
+++ OUTPUT/testing/pluto/crossing-streams-24-ikev2-delete-connswitch-github-2101/east.console.txt
@@ -27,9 +27,11 @@
 "b" #3: sent IKE_SA_INIT request to 192.1.2.45:UDP/500
 "b" #3: processed IKE_SA_INIT response from 192.1.2.45:UDP/500 {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_512 group=DH19}, initiating IKE_AUTH
 "b" #3: sent IKE_AUTH request to 192.1.2.45:UDP/500 with shared-key-mac and FQDN '@east'; Child SA #4 {ESP <0xESPESP} [192.0.20.0/24===192.0.3.254/32]
-"b" #3: processing IKE_AUTH response from 192.1.2.45:UDP/500 containing SK{IDr,AUTH,SA,TSi,TSr}
-"b" #3: initiator established IKE SA; authenticated peer using authby=secret and FQDN '@west'
-"b" #4: initiator established Child SA using #3; IPsec tunnel [192.0.20.0/24===192.0.3.254/32] {ESP/ESN=>0xESPESP <0xESPESP xfrm=AES_GCM_16_256 DPD=passive}
+"b" #3: IKE SA authentication request rejected by peer: AUTHENTICATION_FAILED
+"b" #3: encountered fatal error in state IKE_AUTH_I
+"b" #4: connection is supposed to remain up; revival attempt 1 scheduled in 0 seconds
+IMPAIR: "b" #4: revival: skip scheduling CONNECTION_REVIVAL event in 0 seconds
+"b" #3: deleting IKE SA (sent IKE_AUTH request)
 east #
  # non-zero counts confirm encrypted traffic flowing
 east #
@@ -40,8 +42,4 @@
  ipsec _kernel state
 east #
  ipsec _kernel policy
-src 192.0.20.0/24 dst 192.0.3.254/32
-       dir out priority PRIORITY ptype main
-       tmpl src 0.0.0.0 dst 0.0.0.0
-               proto esp reqid 0 mode transport
 east # 
--- MAIN/testing/pluto/crossing-streams-24-ikev2-delete-connswitch-github-2101/west.console.txt
+++ OUTPUT/testing/pluto/crossing-streams-24-ikev2-delete-connswitch-github-2101/west.console.txt
@@ -29,7 +29,7 @@
 Connection list:
  
 "a": 192.0.3.253/32===192.1.2.45[@west]...192.1.2.23[@east]===192.0.2.0/24; routed-tunnel; my_ip=192.0.3.253; their_ip=unset;
-"a":   host: oriented; local: 192.1.2.45; remote: 192.1.2.23; established IKE SA: #3;
+"a":   host: oriented; local: 192.1.2.45; remote: 192.1.2.23; established IKE SA: #1;
 "a":   my_updown=ipsec _updown; my_updown-config=noasync,noexec;
 "a":   xauth us:none, xauth them:none, my_username=[any]; their_username=[any]
 "a":   our auth:secret, their auth:secret, our autheap:none, their autheap:none;
@@ -48,11 +48,11 @@
 "a":   sendca: all; our sendcert: always; their sendcert: always;
 "a":   liveness: passive; dpddelay:0s; retransmit-timeout:10s
 "a":   nat-traversal: encapsulation:auto; keepalive:20s
-"a":   routing: routed-tunnel; owner: Child SA #2; established IKE SA: #3; established Child SA: #2;
+"a":   routing: routed-tunnel; owner: Child SA #2; established IKE SA: #1; established Child SA: #2;
 "a":   conn serial: $1;
 "a":   IKEv2 algorithm newest: AES_GCM_16_256-HMAC_SHA2_512-DH19
 "a":   ESP algorithm newest: AES_GCM_16_256; pfsgroup=<Phase1>
-"b": 192.0.3.254/32===192.1.2.45[@west]...192.1.2.23[@east]===192.0.20.0/24; routed-tunnel; my_ip=192.0.3.254; their_ip=unset;
+"b": 192.0.3.254/32===192.1.2.45[@west]...192.1.2.23[@east]===192.0.20.0/24; unrouted; my_ip=192.0.3.254; their_ip=unset;
 "b":   host: oriented; local: 192.1.2.45; remote: 192.1.2.23;
 "b":   my_updown=ipsec _updown; my_updown-config=noasync,noexec;
 "b":   xauth us:none, xauth them:none, my_username=[any]; their_username=[any]
@@ -72,18 +72,15 @@
 "b":   sendca: all; our sendcert: always; their sendcert: always;
 "b":   liveness: passive; dpddelay:0s; retransmit-timeout:10s
 "b":   nat-traversal: encapsulation:auto; keepalive:20s
-"b":   routing: routed-tunnel; owner: Child SA #4; established Child SA: #4;
+"b":   routing: unrouted;
 "b":   conn serial: $2;
-"b":   ESP algorithm newest: AES_GCM_16_256; pfsgroup=<Phase1>
  
-Total IPsec connections: loaded 2, routed 2, active 2
+Total IPsec connections: loaded 2, routed 1, active 1
 west #
  ipsec delete a
 "a": terminating SAs using this connection
 "a" #1: deleting IKE SA (ESTABLISHED_IKE_SA) and sending notification
-"a" #3: deleting IKE SA (ESTABLISHED_IKE_SA) and sending notification
 "a" #2: ESP traffic information: in=0B out=0B
-"b" #4: ESP traffic information: in=0B out=0B
 west #
  # non-zero counts confirm encrypted traffic flowing
```

Finally, this test went pear shaped:
```
--- MAIN/testing/pluto/crossing-streams-20-peer-restarts-ondemand-ike-auth/west.console.txt
+++ OUTPUT/testing/pluto/crossing-streams-20-peer-restarts-ondemand-ike-auth/west.console.txt
@@ -76,48 +76,1013 @@
 IMPAIR: packet from 192.1.2.23:500: blocking inbound message 4
 IMPAIR: start processing inbound drip packet 4
 IMPAIR: stop processing inbound drip packet 4
-"east-west" #4: responder established Child SA using #3; IPsec tunnel [192.0.1.0/24===192.0.2.0/24] {ESP/ESN=>0xESPESP <0xESPESP xfrm=AES_GCM_16_256 DPD=passive}
+timeout waiting 30 seconds for cat /tmp/pluto.log to match #4: responder established Child SA
+output: |    IKEv2 transform type: IKEv2_TRANS_TYPE_KEM (0x4)
```

The-Mule commented:
Ah, I am sorry. Looking, I am now fixing crossing-streams-11-ikev2-permanent-ike-auth-east quickly (fr/om is a typo I make after my testing I think and the second difference I am not sure how happened but I see it now as well, fixing both). 

I am looking at the other failures.

The-Mule commented:
Ad crossing-streams-20-peer-restarts-ondemand-ike-auth:

this is expected and I need to update that test (again, sorry that I missed it):

See, the test expects west to accept east's new connection after the restart and successfully establish it. But because of this PR west now see it still has IKE_AUTH oustanding and let either its own IKE (`#1`) or east's new one suceed based on nonces. Impairing east to have higher nonce will make the test pass (or by setting reject-simultaneous-ike-auth=no on both). I just tested it and will be updateting that. There's one more thing though. Since west IKE was terminated by `terminate_ike_family()` the following will be missing from `ipsec _kernel policy` on the west too:

```
-src 192.1.2.23 dst 192.1.2.45
-       proto esp spi 0xSPISPI reqid REQID mode transport
-       replay-window 0
-       dir in
-       sel src 192.1.2.23/32 dst 192.1.2.45/32
```

But that's also fine.  I still need to take a loot at crossing-streams-24-ikev2-delete-connswitch-github-2101 but so far nothing exploded, thankfully.

The-Mule commented:
hm crossing-streams-24-ikev2-delete-connswitch-github-2101 is passing for me (with this branch rebased to the current main):

```
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.03: stop testing crossing-streams-24-ikev2-delete-connswitch-github-2101 (test 1 of 1) after 47.4 seconds
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.03: single test run; leaving domains running
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.03: start post-mortem crossing-streams-24-ikev2-delete-connswitch-github-2101 (test 1 of 1) at 2026-04-09 12:28:51.448453
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.07: ****** crossing-streams-24-ikev2-delete-connswitch-github-2101 (test 1 of 1) passed ******
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.07: stop post-mortem crossing-streams-24-ikev2-delete-connswitch-github-2101 (test 1 of 1) after 0.3 seconds
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.07: Progress:
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.07:   tests/good/results/passed: 1
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.07:   total: 1
crossing-streams-24-ikev2-delete-connswitch-github-2101 50.07:   total/passed: 1
```

Wasn't it something else not yet in main?

Anyway, the following three should be working fine now:

```
crossing-streams-11-ikev2-permanent-ike-auth-east
crossing-streams-11-ikev2-permanent-ike-auth-west
crossing-streams-20-peer-restarts-ondemand-ike-auth
```

cagney commented:
> hm crossing-streams-24-ikev2-delete-connswitch-github-2101 is passing for me (with this branch rebased to the current main)

I consistently get the opposite :-) :-( I guess it's unstable.

West fully establishes `#1/#1`:
```
"a" #1: initiating IKEv2 connection to 192.1.2.23 using UDP
IMPAIR: "a" #1: suppressing retransmits; scheduling timeout in 10 seconds
"a" #1: sent IKE_SA_INIT request to 192.1.2.23:UDP/500
"a" #1: processed IKE_SA_INIT response from 192.1.2.23:UDP/500 {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_512 group=DH19}, initiating IKE_AUTH
IMPAIR: "a" #1: suppressing retransmits; scheduling timeout in 10 seconds
"a" #1: sent IKE_AUTH request to 192.1.2.23:UDP/500 with shared-key-mac and FQDN '@west'; Child SA #2 {ESP <0x8b185422} [192.0.3.253/32===192.0.2.0/24]
"a" #1: processing IKE_AUTH response from 192.1.2.23:UDP/500 containing SK{IDr,AUTH,SA,TSi,TSr}
"a" #1: initiator established IKE SA; authenticated peer using authby=secret and FQDN '@east'
"a" #2: initiator established Child SA using #1; IPsec tunnel [192.0.3.253/32===192.0.2.0/24] {ESP/ESN=>0x94c99f97 <0x8b185422 xfrm=AES_GCM_16_256 DPD=passive}
```
then east's IKE_SA_INIT arrives, so it isn't crossed.
```
"a" #3: processing IKE_SA_INIT request from 192.1.2.23:UDP/500 containing SA,KE,Ni,N(IKEV2_FRAGMENTATION_SUPPORTED),N(NAT_DETECTION_SOURCE_IP),N(NAT_DETECTION_DESTINATION_IP)
"a" #3: proposal 1:IKE=AES_GCM_16_256-HMAC_SHA2_512-ECP_256 chosen from remote proposals 1:IKE:ENCR=AES_GCM_16_256;PRF=HMAC_SHA2_512;PRF=HMAC_SHA2_256;KEM=ECP_256;KEM=ECP_384;KEM=ECP_521;KEM=CURVE25519;KEM=MODP4096;KEM=MODP3072;KEM=MODP2048;KEM=MODP8192[first-match] 2:IKE:ENCR=AES_GCM_16_128;PRF=HMAC_SHA2_512;PRF=HMAC_SHA2_256;KEM=ECP_256;KEM=ECP_384;KEM=ECP_521;KEM=CURVE25519;KEM=MODP4096;KEM=MODP3072;KEM=MODP2048;KEM=MODP8192 3:IKE:ENCR=CHACHA20_POLY1305;PRF=HMAC_SHA2_512;PRF=HMAC_SHA2_256;KEM=ECP_256;KEM=ECP_384;KEM=ECP_521;KEM=CURVE25519;KEM=MODP4096;KEM=MODP3072;KEM=MODP2048;KEM=MODP8192 4:IKE:ENCR=AES_CBC_256;PRF=HMAC_SHA2_512;PRF=HMAC_SHA2_256;INTEG=HMAC_SHA2_512_256;INTEG=HMAC_SHA2_256_128;KEM=ECP_256;KEM=ECP_384;KEM=ECP_521;KEM=CURVE25519;KEM=MODP4096;KEM=MODP3072;KEM=MODP2048;KEM=MODP8192 5:IKE:ENCR=AES_CBC_128;PRF=HMAC_SHA2_512;PRF=HMAC_SHA2_256;INTEG=HMAC_SHA2_512_256;INTEG=HMAC_SHA2_256_128;KEM=ECP_256;KEM=ECP_384;KEM=ECP_521;KEM=CURVE25519;KEM=MODP4096;KEM=MODP3072;KEM=MODP2048;KEM=MODP8192
"a" #3: sent IKE_SA_INIT response to 192.1.2.23:UDP/500 {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_512 group=DH19}, expecting IKE_AUTH, IKE_INTERMEDIATE, or IKE_AUTH (EAP)
"a" #3: received first encrypted IKE_AUTH message containing SK payload from 192.1.2.23:UDP/500, completing Key Exchange computation (SKEYSEED) in the background
"a" #3: processing IKE_AUTH request from 192.1.2.23:UDP/500 containing SK{IDi,IDr,AUTH,SA,TSi,TSr}
"a" #3: IKE SA #1 recently established
"a" #3: rejecting IKE_AUTH request; IKE SA #1 already established
"a" #3: responding to IKE_AUTH message (ID 1) from 192.1.2.23:500 with encrypted notification AUTHENTICATION_FAILED
"a" #3: encountered fatal error in state IKE_SA_INIT_R
"a" #3: deleting IKE SA (sent IKE_SA_INIT response)
```

The comment, at least, isn't correct.  The last exchange was <1s ago, not IKE_AUTH.

```                /* Was candidate established within the last second of the current IKE_AUTH request arrival? */
                if (candidate->sa.st_state->kind == STATE_V2_ESTABLISHED_IKE_SA) {
                        deltatime_t age = monotime_diff(mononow(), candidate->sa.st_v2_msgid_windows.initiator.last_recv);
                        if (deltatime_cmp(age, <, one_second)) {
                                llog(RC_LOG, ike->sa.logger, "IKE SA "PRI_SO" recently established",
                                                pri_so(candidate->sa.st_serialno));
                                simultaneous_ike = candidate;
                                break;
                        }
                }
```

cagney commented on programs/pluto/ikev2_auth.c:
```diff
@@ -1090,13 +1090,23 @@ lset_t proposed_v2AUTH(struct ike_sa *ike,
-		const struct ike_sa *ike,
-		const struct msg_digest *md)
+struct ike_sa *check_simultaneous_ike_auth(const struct connection *c,
+               const struct ike_sa *ike,
```
I think the const on `ike` should be dropped.  Its the reason for the confusing cast:
```
return (struct ike_sa *) ike;
```

cagney commented on programs/pluto/ikev2_auth.c:
```diff
@@ -1109,35 +1119,57 @@ struct ike_sa *get_sa_with_outstanding_ike_auth_request(const struct connection
+		ldbg(ike->sa.logger, "preferring the simultaneous IKE SA "PRI_SO" over the current IKE SA "PRI_SO,
+				pri_so(simultaneous_ike->sa.st_serialno), pri_so(ike->sa.st_serialno));
+		return (struct ike_sa *) ike;
+  } else {
```
there's something wrong with the indentation of the `} else {`

cagney commented on programs/pluto/ikev2_auth.c:
```diff
@@ -1109,35 +1119,57 @@ struct ike_sa *get_sa_with_outstanding_ike_auth_request(const struct connection
-		if (simultaneous_ike == NULL || simultaneous_ike == ike) {
-			continue;
-		} else if (simultaneous_ike->sa.st_sa_role != SA_INITIATOR) {
+		struct ike_sa *candidate = pexpect_ike_sa(sf.st);
```
for the case where the condition is going to continue/return, code tends to be written with the syle:
```
   if (cond) {
       continue;
   }

   if (cond2) {
       continue;
   }
```
i.e, no `else` and blank line after the continue/return

cagney commented on programs/pluto/ikev2_ike_auth.c:
```diff
@@ -681,41 +681,18 @@ stf_status process_v2_IKE_AUTH_request_standard_payloads(struct ike_sa *ike, str
-			     pri_so(simultaneous_ike->sa.st_serialno));
-			terminate_ike_family(&simultaneous_ike, REASON_SUPERSEDED_BY_NEW_SA, HERE);
-		}
+	struct ike_sa *ike_to_reject = check_simultaneous_ike_auth(c, ike, md);
```
this is happening before IKE is authenticated, yet it appears it can cause the local IKE SA to be torn down?

would it be safer to let the IKE SA establish and then just worry about rejecting the child sa

The-Mule replied:
> I'm worried by an unauthenticated IKE SA being able to delete something locally.

This is a silent local decision (ie. we are not sending DELETE notification). Is there some kind of security problem we might be dealing with? (also please see below since we are limiting all this functionality to the permanent connections).

> would it be safer to let the IKE SA establish and then just worry about rejecting the child sa

It's important to mention that we are here only dealing with permanent connections. If I understand correctly the risk here is hat we are deleting local IKE SA without knowing if the other one leads to successfully authnetication. But  for permanent connections credentials are symmetric, right? So if the other one fails then the delete one would fail too. Or is there some other risk?

The thing is that letting both IKE_AUTH complete might be quite a big hit to the current approach.

cagney reviewed (commented):
I'm worried by an unauthenticated IKE SA being able to delete something locally.

cagney commented:
btw, if the IKE SA is authenticated, TEMPORARY_FAILURE becomes slightly more valid response to the Child SA.

The-Mule commented:
> I consistently get the opposite :-) :-( I guess it's unstable.

Oh, you're right, see - there are two check for IKE_AUTH crossing:

1. ikev2_ike_auth.c - Early, before traffic selectors are look at.
2. ikev2_child.c - Late, after traffic selectors are look at.

And it turns out (as revealed by crossing-streams-24-ikev2-delete-connswitch-github-2101) that early check can detect false positives (especially the "1 second thing"). I actually even think we don't need the early check at all now! If we keep only the second one we should still detect all IKE_AUTH crossings on the same connection (even after connswitch) while removing false positives.

Edit: The reason why I see it passing and you see it failing is that the "1 second check" is breaking it for you when the timing is just right. Obviously, it's false positive.

The-Mule commented:
Right, leaving just the late check passed my testing and it actually revealed a minor issue in my code - `record_v2N_response(v2N_AUTHENTICATION_FAILED)` was sent twice in programs/pluto/ikev2_child.c, there is now:

```
record_v2N_response(ike->sa.logger, ike, md, v2N_AUTHENTICATION_FAILED, empty_shunk, ENCRYPTED_PAYLOAD);
return v2N_AUTHENTICATION_FAILED;
```
I can drop `record_v2N_response()` because it will be called upon returning `v2N_AUTHENTICATION_FAILED` anyway.

Let me think about removing "early" check a bit more tomorrow just to make sure I am not missing anything and then I'll fix formatting errors too and update PR, leaving just one check (after traffic selectors are handled) makes thing simpler.

The-Mule commented:
I removed the "early" check ([01a8a79](https://github.com/libreswan/libreswan/pull/2447/commits/01a8a799573fd04d7d26f8cfc01169a1a349876a)), removed duplicated AUTHENTICATION_FAILED notification ([1eb1404](https://github.com/libreswan/libreswan/pull/2447/commits/1eb1404437343c8b3eb1e8794b01fead9a787a9b)) and fixed formatting ([bddae82](https://github.com/libreswan/libreswan/pull/2447/commits/bddae82a958968fdfcb88fe8744a6516a58b6cd9)). It still fixes my problem and crossing-streams tests keep working for me:

```
# ./kvm results | grep -v untested
results_command /root/libreswan/testing
testing/pluto/crossing-streams-03-ikev2-delete-ike passed
testing/pluto/crossing-streams-03-ikev2-delete-child passed
testing/pluto/crossing-streams-03-ikev2-delete-ike-vs-child passed
testing/pluto/crossing-streams-03-ikev2-delete-child-vs-ike passed
testing/pluto/crossing-streams-04-ikev2-initiator-child-up-ike-down passed
testing/pluto/crossing-streams-11-ikev2-permanent-ike-auth-east passed
testing/pluto/crossing-streams-11-ikev2-permanent-ike-auth-west passed
testing/pluto/crossing-streams-20-peer-restarts-ondemand-ike-sa-init passed
testing/pluto/crossing-streams-20-peer-restarts-ondemand-ike-auth passed
testing/pluto/crossing-streams-21-second-child-crypto-ikev2 failed east:output-different west:output-different
testing/pluto/crossing-streams-23-conswitch-ike-sa-init-ikev2 passed
testing/pluto/crossing-streams-24-ikev2-delete-connswitch-github-2101 passed

Summary:
  stats/good/ignored/platform=alpine!=linux: 4
  stats/good/ignored/platform=debian!=linux: 1
  stats/good/ignored/platform=fedora!=linux: 4
  stats/good/ignored/platform=freebsd!=linux: 8
  stats/good/ignored/platform=netbsd!=linux: 33
  stats/good/ignored/platform=openbsd!=linux: 6
  stats/skip/ignored/status=skip!=good: 1
  stats/skiptest/ignored/status=skiptest!=good: 12
  stats/wip/ignored/status=wip!=good: 162
  tests/good/errors/output-different: 1
  tests/good/results/failed: 1
  tests/good/results/passed: 11
  tests/good/results/untested: 1173
  tests/skip/results/untested: 1
  tests/skiptest/results/untested: 12
  tests/wip/results/untested: 162
  total: 1360
  total/failed: 1
  total/ignored: 231
  total/passed: 11
  total/untested: 1117
  ```
  
  Test `crossing-streams-21-second-child-crypto-ikev2` fails but the failure is irrelevant to these changes (ie. it fail on `main` too). I believe that removing early check should solve problem you saw with `crossing-streams-24-ikev2-delete-connswitch-github-2101`.

cagney commented:
@The-Mule results are clean, did you want to cleanup the commits before I merge?

The-Mule commented:
Perfect, I squashed the last fixups to previous commits and rebase onto current main. I don't have access to my testbed right now but I'll try to re-test ASAP.

The-Mule commented:
Alright, retested after the rebase and everything seems to be in order on my test machine:

```
 ****** crossing-streams-03-ikev2-delete-ike (test 1 of 12) passed ******
 ****** crossing-streams-03-ikev2-delete-child (test 2 of 12) passed ******
 ****** crossing-streams-03-ikev2-delete-ike-vs-child (test 3 of 12) passed ******
 ****** crossing-streams-03-ikev2-delete-child-vs-ike (test 4 of 12) passed ******
 ****** crossing-streams-04-ikev2-initiator-child-up-ike-down (test 5 of 12) passed ******
 ****** crossing-streams-11-ikev2-permanent-ike-auth-east (test 6 of 12) passed ******
 ****** crossing-streams-11-ikev2-permanent-ike-auth-west (test 7 of 12) passed ******
 ****** crossing-streams-20-peer-restarts-ondemand-ike-sa-init (test 8 of 12) passed ******
 ****** crossing-streams-20-peer-restarts-ondemand-ike-auth (test 9 of 12) passed ******
 ****** crossing-streams-21-second-child-crypto-ikev2 (test 10 of 12) passed ******
 ****** crossing-streams-23-conswitch-ike-sa-init-ikev2 (test 11 of 12) passed ******
 ****** crossing-streams-24-ikev2-delete-connswitch-github-2101 (test 12 of 12) passed ******
 ```

cagney reviewed (changes requested):
@The-Mule can this please be reduced to the following commits:

- add the impair
- change IKE_AUTH
- update tests

When I diff the top of tree to mainline I see far fewer changes.

I thinking for now the default be off; sending back AUTHENTICATION_FAILED for something that is closer to TEMPORARY_FAILURE isn't right (I tried sending back TEMPORARY_FAILURE, it didn't go so well, so next round).

The-Mule commented:
Done. 

> I thinking for now the default be off; sending back AUTHENTICATION_FAILED for something that is closer to TEMPORARY_FAILURE isn't right (I tried sending back TEMPORARY_FAILURE, it didn't go so well, so next round).

The reason why I switched from TEMPORARY_FAILURE to AUTHENTICATION_FAILED is that I needed "fatal" result on the other end so that the remote IKE SA is torn down (and revived if needed) but TEMPORARY_FAILURE is non-fatal. But yes, this is perhaps something that can be addressed later.

cagney commented:
> The reason why I switched from TEMPORARY_FAILURE to AUTHENTICATION_FAILED is that I needed "fatal" result on the other end so that the remote IKE SA is torn down (and revived if needed) but TEMPORARY_FAILURE is non-fatal. But yes, this is perhaps something that can be addressed later.

I know.  Returning TEMPORARY_FAILURE assumes a co-operative peer.

Technically, we're meant to cook up a new error and:
_Extension documents may define new error notifications with these semantics, but MUST NOT use them unless the peer has been shown to understand them, such as by using the Vendor ID payload._

since we're re-defining AUTHENTICATION_FAILED.

Hence, having it disabled by default.

The-Mule commented:
> Hence, having it disabled by default.

Sounds good to me (it has a very specific use case and hence I think making it non-default behaviour is fine), but that means we can revive crossing-stream-11-* now since they again makes sense. Also, I need add option to make the two new tests working.  Shall I add a commit changing the default from 'yes' to 'no'?

cagney reviewed (changes requested):
(now I can see the changes in full, thanks).

Just some text shuffling.

- could you copy the change log to the PR's first entry; that way I can pretty much verbatim copy it into the merge commit, I'll be --author-ing the merge so your name will be on it.

- can you also expand the comment before  `check_simultaneous_ike_auth()` to include similar information  - this way someone trying to understand the code will find it; it should also mention the rationale for AUTHENTICATION_FAILED and not TEMPORARY_FAILURE

- I think there should be a comment, perhaps near the hunk_cmp() call, pointing out that because this is detecting a crossing Child SA during IKE AUTH, it needs to compare the IKE SA's nuance (there's no CREATE_CHILD_SA exchange, and hence, no Child SA nonce to compare)


You might then find that this commit message can be trimmed back so it focuses more on what was changed.

cagney commented:
>  Shall I add a commit changing the default from 'yes' to 'no'?

Yes please.

The-Mule committed:
Disabled reject-simultaneous-ike-auth by default

Signed-off-by: Ondrej Moris <omoris@redhat.com>

The-Mule committed:
testing: tests for reject_simultaneous_ike_auth

Two new tests to check simultaneous IKE_AUTH requests handling when
reject_simultaneous_ike_auth option enabled.

Signed-off-by: Ondrej Moris <omoris@redhat.com>

The-Mule commented:
I changed the default of reject-simultaneous-ike-auth from 'yes' to 'no', updated two new tests to enable the option and restored back all previous crossing-stream-11-* tests and reverted one more test change that was needed becaouse of the default. IOW, now the only diff in testing/ between this PR and main should be two new tests and nothing else. 

I still need to do text shuffling though :).

The-Mule commented:
Alright I tried to update the comments as well as PR description (it now matches main commit), also retested reverted test changes and it looks good:

```
 ****** crossing-streams-03-ikev2-delete-ike (test 1 of 16) passed ******
 ****** crossing-streams-03-ikev2-delete-child (test 2 of 16) passed ******
 ****** crossing-streams-03-ikev2-delete-ike-vs-child (test 3 of 16) passed ******
 ****** crossing-streams-03-ikev2-delete-child-vs-ike (test 4 of 16) passed ******
 ****** crossing-streams-04-ikev2-initiator-child-up-ike-down (test 5 of 16) passed ******
 ****** crossing-streams-11-ewwew-ikev2-permanent (test 6 of 16) passed ******
 ****** crossing-streams-11-ewwwe-ikev2-permanent (test 7 of 16) passed ******
 ****** crossing-streams-11-wewew-ikev2-permanent (test 8 of 16) passed ******
 ****** crossing-streams-11-wewwe-ikev2-permanent (test 9 of 16) passed ******
 ****** crossing-streams-12-ikev2-permanent-ike-auth-east (test 10 of 16) passed ******
 ****** crossing-streams-12-ikev2-permanent-ike-auth-west (test 11 of 16) passed ******
 ****** crossing-streams-20-peer-restarts-ondemand-ike-sa-init (test 12 of 16) passed ******
 ****** crossing-streams-20-peer-restarts-ondemand-ike-auth (test 13 of 16) passed ******
 ****** crossing-streams-21-second-child-crypto-ikev2 (test 14 of 16) passed ******
 ****** crossing-streams-23-conswitch-ike-sa-init-ikev2 (test 15 of 16) passed ******
 ****** crossing-streams-24-ikev2-delete-connswitch-github-2101 (test 16 of 16) passed ******
```

cagney reviewed (approved):
sorry, should have merged this earlier
